src/api/app/models/staging/staging_project.rb

```diff
@@ -0,0 +1,141 @@
+    has_many :status_reports, through: :repositories, inverse_of: :checkable
+    belongs_to :staging_workflow
+
+    def staging_identifier
+      name[/.*:Staging:(.*)/, 1]
+    end
+
+    def untracked_requests
+      requests_to_review - staged_requests
+    end
+
+    # The difference between staged requests and requests to review is that staged requests are assigned to the staging project.
+    # The requests to review are requests which are not related to the staging project (unless they are also staged).
+    # They simply need a review from the maintainers of the staging project.
+    def requests_to_review
+      @requests_to_review ||= BsRequest.with_open_reviews_for(by_project: name)
+    end
+
+    def building_repositories
+      set_buildinfo unless @building_repositories
+      @building_repositories
+    end
+
+    def broken_packages
+      set_buildinfo unless @broken_packages
+      @broken_packages
+    end
+
+    def missing_reviews
+      return @missing_reviews if @missing_reviews
+
+      @missing_reviews = []
+      attribs = [:by_group, :by_user, :by_project, :by_package]
+
+      (requests_to_review + staged_requests).uniq.each do |request|
+        request.reviews.each do |review|
+          next if review.state == :accepted || review.by_project == name
+          # FIXME: this loop (and the inner if) would not be needed
+          # if every review only has one valid by_xxx.
+          # I'm keeping it to mimic the python implementation.
+          # Instead, we could have something like
+          # who = review.by_group || review.by_user || review.by_project || review.by_package
+          attribs.each do |att|
+            who = review.send(att)
+            if who
+              @missing_reviews << { id: review.id, request: request.number, state: review.state.to_s, package: request.first_target_package, by: who }
+            end
+          end
+        end
+      end
+
+      @missing_reviews
+    end
+
+    def overall_state
+      @overall_state ||= state
+    end
+
+    private
+
+    def state
+      return :empty unless staged_requests.exists?
+      return :unacceptable if untracked_requests.present? || staged_requests.obsolete.present?
+
+      overall_state = build_state
+      overall_state = check_state if overall_state == :acceptable
+
+      return :review if overall_state == :acceptable && missing_reviews.present?
+
+      overall_state
+    end
+
+    def build_state
+      set_buildinfo
+
+      return :building if building_repositories.present?
+      return :failed if broken_packages.present?
+
+      :acceptable
+    end
+
+    def relevant_status_reports
+      @relevant_status_reports ||= status_reports.where(uuid: repositories.map(&:build_id))
+    end
+
+    def missing_checks?
+      relevant_status_reports.any? { |report| report.missing_checks.present? }
+    end
+
+    def check_state
```

Comment:
Why not:

```
def check_state
  status_checks = Status::Check.where(status_reports_id: relevant_status_reports)
  if missing_checks? || status_checks.pending.exists?
    :testing
  elsif status_checks.failed.exists?
    :failed
  else
    :acceptable
  end
end

```

Reply:
We had to pick one. And since there is a slight preference within the team to use guard clauses, we picked that one.

Reply:
@vpereira option is repeating some code `status_checks.failed.exists?`. I also find the other (original) more readable.

Reply:
@vpereira: Same as above, this is a common pattern -> https://refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html

Reply:
yes guard clauses are really common, however i dont see a nested conditional here.. interesting that codeclimate do not complain. SonarQube uses the ratio between number of lines x number of returns in the cyclomatic complexity calculation..

Reply:
> @vpereira option is repeating some code `status_checks.failed.exists?`. I also find the other (original) more readable.

@Ana06 where is the repeated code? The conditions are the same, just instead of return, return, if, elsif, else..

Reply:
@coolo its plain taste, yes thats the only valid argument here. No nested conditional, the code - in a 4 lines of code method - with if/elseif/else isnt less readable.. is just taste. Thats why I`m just  commenting and not asking for change.

Reply:
So you agree that you have no argument for not using guard clauses and this discussion is not going anywhere?:-)

Reply:
a suggestion don't have to go anywhere. Don't take it personal, to write perl looking ruby is quite ok here ;)

Reply:
btw the codecoverage handles as bad smell 4 or more return-statements per method.